### PR TITLE
Treat grammar warnings as non-fatal

### PR DESCRIPTION
**Pull request by lildude**
## Description

We have a very long list of outstanding grammar issues in https://github.com/github/linguist/issues/3924 which isn't getting much smaller,  mostly because of the number of non-fatal issues that really aren't much of an issue. This output is what is produced when I update the grammars with each release and it makes it really hard to spot new legitimate issues.

This PR addresses one of @Alhadis's longest requests:

> Could we downgrade "Missing include in grammar:… the scope cannot be found" errors to build-time warnings, please?

If a contributor attempts to add or replace a grammar with legitimate errors, they get a failure as before:

<details>
<summary>Example output for adding/replacing grammar with fatal errors</summary>

```console
$ script/add-grammar --replace turtle.tmbundle https://github.com/peta/turtle.tmbundle
Checking docker is installed and running
$ docker ps
Deregistering: vendor/grammars/turtle.tmbundle
$ git submodule deinit vendor/grammars/turtle.tmbundle
$ git rm -rf vendor/grammars/turtle.tmbundle
$ script/grammar-compiler update -f
Registering new submodule: vendor/grammars/turtle.tmbundle
$ git submodule add -f https://github.com/peta/turtle.tmbundle vendor/grammars/turtle.tmbundle
$ script/grammar-compiler add vendor/grammars/turtle.tmbundle
  > 
  > 5 errors found in new grammar 'repository `vendor/grammars/turtle.tmbundle` (from https://github.com/peta/turtle.tmbundle)':
  > - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)                          (?<PNAME_NS> (?: (?: [\`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 68))
  > - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)((?<=\s|^|_)(?:[\p{L}\p{M}] `...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 57))
  > - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)                  (?<PN_CHARS_U>[\p{L}\p{M`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 73))
  > - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`\[[\u20\u9\uD\uA]*\]`": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 4))
  > - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)(                         (?: [\p{L}\p{M}] | [:0`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 121))
  > 
  > Compilation failed. Aborting

$
```

</details>

If they attempt to add or replace a grammar with only warnings, they don't get a failure, but they do get a warning:

<details>
<summary>Example output for adding/replacing grammar with non-fatal warnings</summary>

```console
$ script/add-grammar --replace vscode-lean https://github.com/leanprover/vscode-lean
Checking docker is installed and running
$ docker ps
Deregistering: vendor/grammars/vscode-lean
$ git submodule deinit vendor/grammars/vscode-lean
$ git rm -rf vendor/grammars/vscode-lean
$ script/grammar-compiler update -f
Registering new submodule: vendor/grammars/vscode-lean
$ git submodule add -f https://github.com/leanprover/vscode-lean vendor/grammars/vscode-lean
$ script/grammar-compiler add vendor/grammars/vscode-lean
  > 7 warnings found when compiling new grammar 'repository `vendor/grammars/vscode-lean` (from https://github.com/leanprover/vscode-lean)':
  > - Unknown keys in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) contains invalid keys (`Repository[lists].Patterns[0].why`, `Repository[lists].Patterns[1].why`, `Repository[raw_block_disabled].why`, `information_for_contributors`, `version`)
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.html.derivative` but the scope cannot be found
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `source.json.comments` but the scope cannot be found
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `source.asp.vb.net` but the scope cannot be found
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.git-commit` but the scope cannot be found
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.git-rebase` but the scope cannot be found
  > - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.log` but the scope cannot be found
  > 
  > These warnings are not fatal, but may mean the syntax highlighting on GitHub.com may not be as expected.
  > 
Caching grammar license
$ bundle exec licensed cache -c vendor/licenses/config.yml
Updating grammar documentation in vendor/README.md
$ bundle exec rake samples
$ script/sort-submodules
$ script/list-grammars

$
```

</details>

I've also hidden these warnings by default when updating/compiling all the grammars:

<details>
<summary>Example output for updating/compiling all grammars</summary>

```console
$ script/grammar-compiler update -f
 378 / 378  100.00% 5m36s
done! processed 378 grammars

- [ ] repository `vendor/grammars/abap.tmbundle` (from https://github.com/pvl/abap.tmbundle) (1 errors)
  - Invalid regex in grammar: `source.abap` (in `Syntaxes/ABAP.tmLanguage`) contains a malformed regex (regex "`(?i)(?<=(?:^|\s|~|->|-|=>))([a-z`...": lookbehind assertion is not fixed length (at offset 26))

- [ ] repository `vendor/grammars/atom-language-perl6` (from https://github.com/perl6/atom-language-perl6) (6 errors)
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x) ( [\p{Digit}\p{Alpha}'\-_]+`...": unknown property name after \P or \p (at offset 16))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`[\p{Digit}\p{Alpha}'\-_]+`": unknown property name after \P or \p (at offset 9))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x)(\$|@|%|&)(\.|\*|:|!|\^|~|`...": unknown property name after \P or \p (at offset 57))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x)(?<!\\)(\$|@|%|&)(?!\$)(`...": unknown property name after \P or \p (at offset 75))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(`": missing ) (at offset 1))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`)`": unmatched parentheses (at offset 0))

- [ ] repository `vendor/grammars/c.tmbundle` (from https://github.com/textmate/c.tmbundle) (4 errors)
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:APNot(?:CreatedErr|Foun`...": definition too long (282084 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:E(?:A(?:ddressDesc|rray`...": definition too long (52248 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:CATransform3DIdentity|KERNE`...": definition too long (33340 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`(\s*)(\b(?:A(?:E(?:Build(?:Apple`...": definition too long (58589 bytes))

- [ ] repository `vendor/grammars/gap-tmbundle` (from https://github.com/dhowden/gap-tmbundle) (3 errors)
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(16Bits_AssocWord|16Bits_Depth`...": definition too long (65523 bytes))
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(IndicesChiefNormalSteps|Indic`...": definition too long (65529 bytes))
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(SMTX_GoodElementGModule|SMTX_`...": definition too long (42470 bytes))

- [ ] repository `vendor/grammars/language-jsonnet` (from https://github.com/google/language-jsonnet) (1 errors)
  - Invalid regex in grammar: `source.jsonnet` (in `grammars/jsonnet.cson`) contains a malformed regex (regex "`\b[a-zA-Z_][a-z0-9A-Z_]*\s*(\([^`...": nothing to repeat (at offset 43))

- [ ] repository `vendor/grammars/mathematica-tmbundle` (from https://github.com/shadanan/mathematica-tmbundle) (1 errors)
  - Invalid regex in grammar: `source.mathematica` (in `Syntaxes/Mathematica.tmLanguage`) contains a malformed regex (regex "`(\b|(?<=_))(Abort|AbortKernels|A`...": definition too long (54020 bytes))

- [ ] repository `vendor/grammars/objective-c.tmbundle` (from https://github.com/textmate/objective-c.tmbundle) (2 errors)
  - Invalid regex in grammar: `source.objc.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:AB(?:AddRecordsError|Multip`...": definition too long (32854 bytes))
  - Invalid regex in grammar: `source.objc.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:M(?:Action(?:A(?:pplica`...": definition too long (44404 bytes))

- [ ] repository `vendor/grammars/sublime-autoit` (from https://github.com/AutoIt/SublimeAutoItScript) (2 errors)
  - Invalid regex in grammar: `source.autoit` (in `AutoIt.tmLanguage`) contains a malformed regex (regex "`\b(?i:_array1dtohistogram|_array`...": definition too long (39591 bytes))
  - Invalid regex in grammar: `source.autoit` (in `AutoIt.tmLanguage`) contains a malformed regex (regex "`\b(?i:_guictrltoolbar_getbuttoni`...": definition too long (39600 bytes))

- [ ] repository `vendor/grammars/turtle.tmbundle` (from https://github.com/peta/turtle.tmbundle) (5 errors)
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)  (?<PNAME_NS> (?: (?: [\`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 68))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)( (?: [\p{L}\p{M}] | [:0`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 121))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)  (?<PN_CHARS_U>[\p{L}\p{M`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 73))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)((?<=\s|^|_)(?:[\p{L}\p{M}] `...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 57))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`\[[\u20\u9\uD\uA]*\]`": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 4))

The grammar library contains 25 errors

$
```

</details>

This smaller output will be what is used to continue to track problems with the grammars in https://github.com/github/linguist/issues/3924 .

These warnings can however be revealed with the `-v` flag:

<details>
<summary>Example verbose output for updating/compiling all grammars</summary>

```console
$ script/grammar-compiler update -f -v
 378 / 378  100.00% 3m56s
done! processed 378 grammars

- [ ] repository `vendor/grammars/NimLime` (from https://github.com/nim-lang/NimLime) (2 errors)
  - Missing include in grammar: `source.nim` (in `Syntaxes/Nim.tmLanguage`) attempts to include `source.asm` but the scope cannot be found
  - Missing include in grammar: `source.nim` (in `Syntaxes/Nim.tmLanguage`) attempts to include `text.html.markdown.multimarkdown` but the scope cannot be found

- [ ] repository `vendor/grammars/Scalate.tmbundle` (from https://github.com/scalate/Scalate.tmbundle) (1 errors)
  - Missing include in grammar: `source.scaml` (in `Syntaxes/Scaml.tmLanguage`) attempts to include `source.js.jquery` but the scope cannot be found

- [ ] repository `vendor/grammars/Sublime-SQF-Language` (from https://github.com/JonBons/Sublime-SQF-Language) (1 errors)
  - Unknown keys in grammar: `source.sqf` (in `SQF.tmLanguage`) contains invalid keys (`Patterns[20].block`)

- [ ] repository `vendor/grammars/SystemVerilog` (from https://github.com/TheClams/SystemVerilog) (1 errors)
  - Unknown keys in grammar: `source.systemverilog` (in `SystemVerilog.tmLanguage`) contains invalid keys (`hidden`)

- [ ] repository `vendor/grammars/abap.tmbundle` (from https://github.com/pvl/abap.tmbundle) (1 errors)
  - Invalid regex in grammar: `source.abap` (in `Syntaxes/ABAP.tmLanguage`) contains a malformed regex (regex "`(?i)(?<=(?:^|\s|~|->|-|=>))([a-z`...": lookbehind assertion is not fixed length (at offset 26))

- [ ] repository `vendor/grammars/astro` (from https://github.com/snowpackjs/astro) (7 errors)
  - Unknown keys in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) contains invalid keys (`information_for_contributors`)
  - Unknown keys in grammar: `text.html.astro` (in `tools/vscode/syntaxes/astro.tmLanguage.json`) contains invalid keys (`Patterns[9].Patterns[1].Patterns[0].__DEFAULT_STYLE_INCLUDE_END__`, `Patterns[9].Patterns[1].Patterns[0].__DEFAULT_STYLE_INCLUDE_START__`, `Patterns[9].__DEFAULT_STYLE_NAME_END__`, `Patterns[9].__DEFAULT_STYLE_NAME_START__`)
  - Missing include in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) attempts to include `source.json.comments` but the scope cannot be found
  - Missing include in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) attempts to include `text.git-commit` but the scope cannot be found
  - Missing include in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) attempts to include `text.git-rebase` but the scope cannot be found
  - Missing include in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) attempts to include `text.log` but the scope cannot be found
  - Missing include in grammar: `text.html.markdown.astro` (in `tools/vscode/syntaxes/astro-markdown.tmLanguage.json`) attempts to include `source.asp.vb.net` but the scope cannot be found

- [ ] repository `vendor/grammars/atom-language-perl6` (from https://github.com/perl6/atom-language-perl6) (6 errors)
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x) ( [\p{Digit}\p{Alpha}'\-_]+`...": unknown property name after \P or \p (at offset 16))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`[\p{Digit}\p{Alpha}'\-_]+`": unknown property name after \P or \p (at offset 9))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x)(?<!\\)(\$|@|%|&)(?!\$)(`...": unknown property name after \P or \p (at offset 75))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(`": missing ) (at offset 1))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`)`": unmatched parentheses (at offset 0))
  - Invalid regex in grammar: `source.raku` (in `grammars/raku.cson`) contains a malformed regex (regex "`(?x)(\$|@|%|&)(\.|\*|:|!|\^|~|`...": unknown property name after \P or \p (at offset 57))

- [ ] repository `vendor/grammars/ballerina-grammar` (from https://github.com/ballerina-platform/ballerina-grammar) (1 errors)
  - Unknown keys in grammar: `source.ballerina` (in `syntaxes/ballerina.tmLanguage`) contains invalid keys (`tmlVariables`, `monarchVariables`)

- [ ] repository `vendor/grammars/c.tmbundle` (from https://github.com/textmate/c.tmbundle) (4 errors)
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:APNot(?:CreatedErr|Foun`...": definition too long (282084 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:E(?:A(?:ddressDesc|rray`...": definition too long (52248 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:CATransform3DIdentity|KERNE`...": definition too long (33340 bytes))
  - Invalid regex in grammar: `source.c.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`(\s*)(\b(?:A(?:E(?:Build(?:Apple`...": definition too long (58589 bytes))

- [ ] repository `vendor/grammars/gap-tmbundle` (from https://github.com/dhowden/gap-tmbundle) (3 errors)
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(16Bits_AssocWord|16Bits_Depth`...": definition too long (65523 bytes))
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(IndicesChiefNormalSteps|Indic`...": definition too long (65529 bytes))
  - Invalid regex in grammar: `source.gap` (in `Syntaxes/GAP.tmLanguage`) contains a malformed regex (regex "`\b(SMTX_GoodElementGModule|SMTX_`...": definition too long (42470 bytes))

- [ ] repository `vendor/grammars/haxe-TmLanguage` (from https://github.com/vshaxe/haxe-TmLanguage) (1 errors)
  - Missing include in grammar: `documentation.markdown.injection.haxe` (in `helper/haxe-markdown-docs.tmLanguage`) attempts to include `text.html.markdown` but the scope cannot be found

- [ ] repository `vendor/grammars/language-crystal` (from https://github.com/atom-crystal/language-crystal) (1 errors)
  - Missing include in grammar: `source.crystal` (in `grammars/crystal.cson`) attempts to include `source.js.jquery` but the scope cannot be found

- [ ] repository `vendor/grammars/language-fontforge` (from https://github.com/Alhadis/language-fontforge) (1 errors)
  - Unknown keys in grammar: `source.afm` (in `grammars/afm.cson`) contains invalid keys (`foldStartPattern`, `foldEndPattern`)

- [ ] repository `vendor/grammars/language-haml` (from https://github.com/ezekg/language-haml) (1 errors)
  - Missing include in grammar: `text.haml` (in `grammars/ruby haml.cson`) attempts to include `source.ruby.rails` but the scope cannot be found

- [ ] repository `vendor/grammars/language-jsonnet` (from https://github.com/google/language-jsonnet) (1 errors)
  - Invalid regex in grammar: `source.jsonnet` (in `grammars/jsonnet.cson`) contains a malformed regex (regex "`\b[a-zA-Z_][a-z0-9A-Z_]*\s*(\([^`...": nothing to repeat (at offset 43))

- [ ] repository `vendor/grammars/language-ruby` (from https://github.com/atom/language-ruby) (1 errors)
  - Missing include in grammar: `source.ruby` (in `grammars/ruby.cson`) attempts to include `source.js.jquery` but the scope cannot be found

- [ ] repository `vendor/grammars/language-sass` (from https://github.com/atom/language-sass) (1 errors)
  - Missing include in grammar: `source.sassdoc` (in `grammars/sassdoc.cson`) attempts to include `source.html` but the scope cannot be found

- [ ] repository `vendor/grammars/language-shellscript` (from https://github.com/atom/language-shellscript) (1 errors)
  - Missing include in grammar: `source.shell` (in `grammars/shell-unix-bash.cson`) attempts to include `text.html.textile` but the scope cannot be found

- [ ] repository `vendor/grammars/language-yaml` (from https://github.com/atom/language-yaml) (1 errors)
  - Missing include in grammar: `source.yaml` (in `grammars/yaml.cson`) attempts to include `source.ruby.rails` but the scope cannot be found

- [ ] repository `vendor/grammars/mako-tmbundle` (from https://github.com/marconi/mako-tmbundle) (1 errors)
  - Missing include in grammar: `text.html.mako` (in `Syntaxes/HTML (Mako).tmLanguage`) attempts to include `comment.block` but the scope cannot be found

- [ ] repository `vendor/grammars/mathematica-tmbundle` (from https://github.com/shadanan/mathematica-tmbundle) (1 errors)
  - Invalid regex in grammar: `source.mathematica` (in `Syntaxes/Mathematica.tmLanguage`) contains a malformed regex (regex "`(\b|(?<=_))(Abort|AbortKernels|A`...": definition too long (54020 bytes))

- [ ] repository `vendor/grammars/objective-c.tmbundle` (from https://github.com/textmate/objective-c.tmbundle) (2 errors)
  - Invalid regex in grammar: `source.objc.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:AB(?:AddRecordsError|Multip`...": definition too long (32854 bytes))
  - Invalid regex in grammar: `source.objc.platform` (in `Syntaxes/Platform.tmLanguage`) contains a malformed regex (regex "`\b(?:A(?:M(?:Action(?:A(?:pplica`...": definition too long (44404 bytes))

- [ ] repository `vendor/grammars/ruby-slim.tmbundle` (from https://github.com/slim-template/ruby-slim.tmbundle) (1 errors)
  - Missing include in grammar: `text.slim` (in `Syntaxes/Ruby Slim.tmLanguage`) attempts to include `source.ruby.rails` but the scope cannot be found

- [ ] repository `vendor/grammars/sublime-autoit` (from https://github.com/AutoIt/SublimeAutoItScript) (2 errors)
  - Invalid regex in grammar: `source.autoit` (in `AutoIt.tmLanguage`) contains a malformed regex (regex "`\b(?i:_array1dtohistogram|_array`...": definition too long (39591 bytes))
  - Invalid regex in grammar: `source.autoit` (in `AutoIt.tmLanguage`) contains a malformed regex (regex "`\b(?i:_guictrltoolbar_getbuttoni`...": definition too long (39600 bytes))

- [ ] repository `vendor/grammars/turtle.tmbundle` (from https://github.com/peta/turtle.tmbundle) (5 errors)
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)  (?<PNAME_NS> (?: (?: [\`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 68))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)( (?: [\p{L}\p{M}] | [:0`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 121))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)  (?<PN_CHARS_U>[\p{L}\p{M`...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 73))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`\[[\u20\u9\uD\uA]*\]`": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 4))
  - Invalid regex in grammar: `source.turtle` (in `Syntaxes/Turtle.tmLanguage`) contains a malformed regex (regex "`(?x)((?<=\s|^|_)(?:[\p{L}\p{M}] `...": PCRE does not support \L, \l, \N{name}, \U, or \u (at offset 57))

- [ ] repository `vendor/grammars/vscode-lean` (from https://github.com/leanprover/vscode-lean) (7 errors)
  - Unknown keys in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) contains invalid keys (`Repository[lists].Patterns[0].why`, `Repository[lists].Patterns[1].why`, `Repository[raw_block_disabled].why`, `version`, `information_for_contributors`)
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.html.derivative` but the scope cannot be found
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.log` but the scope cannot be found
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.git-commit` but the scope cannot be found
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `source.json.comments` but the scope cannot be found
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `text.git-rebase` but the scope cannot be found
  - Missing include in grammar: `source.lean.markdown` (in `syntaxes/lean-markdown.json`) attempts to include `source.asp.vb.net` but the scope cannot be found
 
$
```

</details>

I've kept the ability to view these warnings for situations where a grammar author may question why the syntax highlighting on GitHub.com doesn't match what they see in something like VS Code.

As part of this PR, I've also converted the compiler to using Go modules, updated the build env for it, and updated `urfave/cli` to v2 so we can take advantage of default values which needed a few other tweaks due to a change in the API.

I'll update the Docker container used to compile the grammars once this PR has been merged.

## Checklist:

- [x] **I am adding new or changing current functionality**
  
  - [ ] I have added or updated the tests for the new or changed functionality.
    The compiler needs a lot more tests, but that's an effort for another day.

### CODE REVIEW

File: script/add-grammar [hunk 1]
```diff
@@ -43,16 +43,24 @@ def log(msg)
   puts msg if $verbose
 end
 
-def command(*args)
+def command(*args, show_warnings: true)
   log "$ #{args.join(' ')}"
-  output, status = Open3.capture2e(*args)
-  if !status.success?
+  stdout, stderr, status = Open3.capture3(*args)
```
Comment (Alhadis) [inline script/add-grammar:48]: UX nit: it's generally not a good idea to interleave STDERR and STDOUT unless the user is expected to do something with the latter (pipe to a pager, write the results to a file, copy it to the system's clipboard, etc).

File: script/add-grammar [hunk 2]
```diff
@@ -43,16 +43,24 @@ def log(msg)
   puts msg if $verbose
 end
 
-def command(*args)
+def command(*args, show_warnings: true)
```
Comment (Alhadis) [inline script/add-grammar:46]: **UX nit:** Rename this to
```suggestion
def command(*args, hide_warnings: false)
```

IMHO, it's bad program design to add boolean options that default to true anyway, as it logically contradicts the purpose of a "switch"-type option (`--verbose` makes more sense than `--no-verbose`, for instance).

File: script/add-grammar [hunk 3]
```diff
@@ -43,16 +43,24 @@ def log(msg)
...
     output.each_line do |line|
       log "  > #{line}"
     end
     warn "Command failed. Aborting."
     exit 1
   end
+  unless stderr.empty?
+    if show_warnings
+      stderr.each_line do |line|
+        log "  > #{line}"
```
Comment (Alhadis) [inline script/add-grammar:60]: Out-of-scope for this PR, but do you think `script/add-grammar` would be better off implemented as a shell-script? I'd be more than willing to rewrite it (complete with options handling and everything), but it wouldn't hurt to know your thoughts on the matter.

Most of this file is just shelling out to external commands, and then echoing the output, which… feels pretty silly now that I think about it. 😕

File: tools/grammars/compiler/converter.go [hunk 4]
```diff
@@ -75,13 +75,35 @@ func (conv *Converter) AddGrammar(source string) error {
...
+		for _, e := range repo.Errors {
+			switch e.(type) {
+			case *MissingIncludeError, *UnknownKeysError:
+				warnings = append(warnings, e)
+			default:
+				errors = append(errors, e)
+			}
+		}
+		if len(errors) > 0 {
+			fmt.Fprintf(os.Stderr, "The new grammar %s contains %d errors:\n",
```
Comment (Alhadis) [inline tools/grammars/compiler/converter.go:90]: Move the important stuff to the beginning of the line, and the part the user already knows to the end:
```suggestion
			fmt.Fprintf(os.Stderr, "%d errors found in new grammar '%s'\n",
```

File: tools/grammars/compiler/converter.go [hunk 5]
```diff
@@ -75,13 +75,35 @@ func (conv *Converter) AddGrammar(source string) error {
...
+				repo, len(errors))
+			for _, err := range errors {
+				fmt.Fprintf(os.Stderr, "    - %s\n", err)
+			}
+			fmt.Fprintf(os.Stderr, "\n")
+			return fmt.Errorf("Failed to compile the given grammar")
```
Comment (Alhadis) [inline tools/grammars/compiler/converter.go:96]:
```suggestion
			return fmt.Errorf("Compilation failed. Aborting")
```

File: tools/grammars/compiler/converter.go [hunk 6]
```diff
@@ -75,13 +75,35 @@ func (conv *Converter) AddGrammar(source string) error {
...
+		}
+		if len(warnings) > 0 {
+			fmt.Fprintf(os.Stderr, "The new grammar %s contains %d warnings:\n",
```
Comment (Alhadis) [inline tools/grammars/compiler/converter.go:99]:
```suggestion
			fmt.Fprintf(os.Stderr, "%d warnings found when compiling new grammar '%s':\n",
```

File: tools/grammars/compiler/converter.go [hunk 7]
```diff
@@ -75,13 +75,35 @@ func (conv *Converter) AddGrammar(source string) error {
...
+				warnings = append(warnings, e)
+			default:
+				errors = append(errors, e)
+			}
+		}
+		if len(errors) > 0 {
+			fmt.Fprintf(os.Stderr, "The new grammar %s contains %d errors:\n",
+				repo, len(errors))
+			for _, err := range errors {
+				fmt.Fprintf(os.Stderr, "    - %s\n", err)
```
Comment (Alhadis) [inline tools/grammars/compiler/converter.go:93]: Personally, I'm not a fan of indenting error messages, as it takes up valuable space in what could be a narrow terminal window. I'd use the format

~~~
[filename]:[line-number]: [problematic-key]: [problem description]
~~~

~~~console
syntaxes/lean-markdown.json:1720: include: scope `text.log` not found
syntaxes/lean-markdown.json:2150: lists: invalid key for field `Repository[lists].Patterns[0]…`
~~~

**Review by Alhadis — CHANGES_REQUESTED** [review]

File: script/add-grammar (hunk 1 above)
Comment (lildude) [reply script/add-grammar:48]: I'm not interleaving here, I'm explicitly splitting them. I merge them again later though 😁

File: script/add-grammar (hunk 2 above)
Comment (lildude) [reply script/add-grammar:46]: Make sense.

File: script/add-grammar (hunk 3 above)
Comment (lildude) [reply script/add-grammar:60]: Most definitely! This whole shell in Ruby was really annoying me.

File: tools/grammars/compiler/converter.go (hunk 7 above)
Comment (lildude) [reply tools/grammars/compiler/converter.go:93]: I'm not a fan either, but deliberately didn't adjust the formatting to reduce the complexity of this PR. Happy to do so if you're happy with a more complex PR.

Comment (Alhadis) [reply tools/grammars/compiler/converter.go:93]: Thing is, users will no doubt be inclined to forward legitimate mistakes to the grammar upstream. Being able to filter out messages by category (such as "missing" grammar scopes), whether visually or through `grep`, will inevitably spell the difference between _"tl;dr, probably okay tho"_ and an _"I should fix that typo"_.

The fact this format is self-sorting is an added bonus, too.

Comment (lildude) [reply tools/grammars/compiler/converter.go:93]: Looking into it, I think this level of information is beyond the scope of this PR. It's going to need a fair amount of refactoring to cater for line number etc. I think we can still remove the indentation though.

Comment (Alhadis) [reply tools/grammars/compiler/converter.go:93]: Fair enough.

File: script/add-grammar (hunk 2 above)
Comment (lildude) [reply script/add-grammar]: Done

File: tools/grammars/compiler/converter.go (hunk 4 above)
Comment (lildude) [reply tools/grammars/compiler/converter.go]: Done

File: tools/grammars/compiler/converter.go (hunk 5 above)
Comment (lildude) [reply tools/grammars/compiler/converter.go]: Done

File: tools/grammars/compiler/converter.go (hunk 6 above)
Comment (lildude) [reply tools/grammars/compiler/converter.go]: Done

File: tools/grammars/compiler/converter.go (hunk 7 above)
Comment (lildude) [reply tools/grammars/compiler/converter.go]: Indentation removed. I've also removed the checkbox from the individual errors in the main report

**Review by Alhadis — APPROVED** [review]

**Pull request merged**